### PR TITLE
Compression for Tensorflow

### PR DESCRIPTION
**Pull request by liuzhe-lz**

### CODE REVIEW

File: src/sdk/pynni/tests/test_compressor_torch.py [hunk 1]
```diff
@@ -52,13 +31,6 @@ def forward(self, x):
         return F.log_softmax(x, dim=1)
 
 
-def tf2(func):
```
Comment (QuanluZhang) [inline src/sdk/pynni/tests/test_compressor_torch.py:55]: no tf test in this pr?

File: examples/model_compress/model_prune_tf.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
+
+    print('start training')
+    model.fit(train_set[0], train_set[1], batch_size=args.batch_size, epochs=args.pretrain_epochs, validation_data=test_set)
+
+    print('start model pruning')
+    optimizer_finetune = tf.keras.optimizers.SGD(learning_rate=0.001, momentum=0.9, decay=1e-4)
+    pruner = create_pruner(model, args.pruner_name)
+    model = pruner.compress()
+    model.compile(optimizer=optimizer_finetune, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
```
Comment (QuanluZhang) [inline examples/model_compress/model_prune_tf.py:70]: why this model need to be compiled twice?

Comment (QuanluZhang) [reply examples/model_compress/model_prune_tf.py:70]: what if this line is not called?

File: src/sdk/pynni/nni/compression/tensorflow/compressor.py [hunk 3]
```diff
@@ -1,204 +1,300 @@
...
+        self.config = None
+
 
 class Compressor:
     """
-    Abstract base TensorFlow compressor
+    Common base class for all compressors.
+
+    This class is designed for other base classes.
+    Algorithms should inherit ``Pruner`` or Quantizer instead.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/tensorflow/compressor.py:55]: Quantizer -> ``Quantizer``

File: src/sdk/pynni/nni/compression/tensorflow/compressor.py [hunk 4]
```diff
@@ -1,204 +1,300 @@
...
-            config['op_types'] = self._expand_config_op_types(config)
-            if layer.type not in config['op_types']:
-                continue
-            if config.get('op_names') and layer.name not in config['op_names']:
-                continue
-            ret = config
-        if ret is None or ret.get('exclude'):
-            return None
-        return ret
+    The compressed model can only execute in eager mode.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/tensorflow/compressor.py:107]: is this statement still true?

File: src/sdk/pynni/nni/compression/tensorflow/compressor.py [hunk 5]
```diff
@@ -1,204 +1,300 @@
...
+        #self.callback = PrunerCallback(self)
 
-    def step(self):
-        """
-        If user want to update mask every step, user can override this method
+    def compress(self):
         """
+        Apply compression on a pre-trained model.
 
+        If you want to prune the model during training, use callback API instead.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/tensorflow/compressor.py:126]: which one is callback API?

File: src/sdk/pynni/nni/compression/tensorflow/compressor.py [hunk 6]
```diff
@@ -1,204 +1,300 @@
...
-            the layer to instrument the mask
-        config : dict
-            the configuration for generating the mask
+        Returns
+        -------
+        dict of (str, tf.Tensor), or None
+            The key is weight ``Variable``'s name. The value is a mask ``Tensor`` of weight's shape and dtype.
+            If a weight's key does not appear in the return value, that weight will not be pruned.
+            Returning ``None`` means the mask is not changed since last time.
+            Weight names are globally unique, e.g. `model/conv_1/kernel:0`.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/tensorflow/compressor.py:156]: very clear!

File: src/sdk/pynni/nni/compression/tensorflow/compressor.py (hunk 5 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/tensorflow/compressor.py:126]: `PrunerCallback` right?

**Comment by QuanluZhang** [comment]
looks good. please add test for it and fix pipeline error

File: src/sdk/pynni/tests/test_compressor_torch.py (hunk 1 above)
Comment (liuzhe-lz) [reply src/sdk/pynni/tests/test_compressor_torch.py:55]: Maybe in bug bash.

File: examples/model_compress/model_prune_tf.py (hunk 2 above)
Comment (liuzhe-lz) [reply examples/model_compress/model_prune_tf.py:70]: `compile` means "attaching optimizer and loss to the model", not "transform graph to static code".

File: src/sdk/pynni/nni/compression/tensorflow/compressor.py (hunk 4 above)
Comment (liuzhe-lz) [reply src/sdk/pynni/nni/compression/tensorflow/compressor.py:107]: Yes, we apply masks by modifying weight variables, which can only be done in eager mode.

File: src/sdk/pynni/tests/test_compressor_torch.py (hunk 1 above)
Comment (QuanluZhang) [reply src/sdk/pynni/tests/test_compressor_torch.py:55]: sure

File: examples/model_compress/model_prune_tf.py (hunk 2 above)
Comment (QuanluZhang) [reply examples/model_compress/model_prune_tf.py:70]: so the performance will be affected if we apply masks on the model? how much performance degradation it would induce?

File: src/sdk/pynni/nni/compression/tensorflow/compressor.py (hunk 4 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/tensorflow/compressor.py:107]: so the performance will be affected if we apply masks on the model? how much performance degradation it would induce?

Comment (liuzhe-lz) [reply src/sdk/pynni/nni/compression/tensorflow/compressor.py:107]: I don't know...

Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/tensorflow/compressor.py:107]: better to check the performance degradation using simple test

File: docs/zh_CN/Compressor/Pruner.md [hunk 7]
```diff
@@ -102,16 +102,6 @@ pruner.compress()
...
-TensorFlow 代码
-```python
-from nni.compression.tensorflow import FPGMPruner
-config_list = [{
-    'sparsity': 0.5,
-    'op_types': ['Conv2D']
-}]
-pruner = FPGMPruner(model, config_list)
-pruner.compress()
-```
```
Comment (QuanluZhang) [inline docs/zh_CN/Compressor/Pruner.md:114]: i think we should keep Chinese doc unchanged, and let @JunweiSUN to update Chinese doc using crowdin

**Comment by tu1258** [comment]
Would you mind telling me why FPGM Pruner is removed? Is former version not work or you guys are work on a new version now? I may have to use filter pruning for CNN while there is almost no tensorflow related toolkit for it. Is there a stable version here?(no need to be latest) Much thanks!

**Comment by QuanluZhang** [comment]
> Would you mind telling me why FPGM Pruner is removed? Is former version not work or you guys are work on a new version now? I may have to use filter pruning for CNN while there is almost no tensorflow related toolkit for it. Is there a stable version here?(no need to be latest) Much thanks!

@tu1258 As model compression framework on NNI has been refactored, we will re-implement the algorithms on the new framework for tensorflow. also, highly recommend external contributors to contribute the compression algorithms.
